python/pyarrow/_s3fs.pyx

```diff
@@ -283,6 +287,7 @@ cdef class S3FileSystem(FileSystem):
 
         options.allow_bucket_creation = allow_bucket_creation
         options.allow_bucket_deletion = allow_bucket_deletion
+        options.stock_retry_strategy = stock_retry_strategy
```

Comment:
Looking into this, it doesn't seem like we should need to modify the C++ code, since the current API will accept the AWS retry strategies. So what if we kept this logic at the Cython level. Something like:

```suggestion
        if stock_retry_strategy == "standard":
            options.retry_strategy = make_shared[AwsClientStandardRetryStrategy]()
        elif stock_retry_strategy == "adaptive":
            options.retry_strategy = make_shared[AwsClientAdaptiveRetryStrategy]()
        elif isinstance(stock_retry_strategy, str):
            raise ValueError(f"Unknown 'stock_retry_strategy': \"{stock_retry_strategy}\"")
```

You would also need to define `AwsClientStandardRetryStrategy` and others inside of `libarrow_fs.pxd`.

Reply:
Thanks Will. This seems so obvious now you mention it. I'll come back with something to this effect.